feat(runLinter): allow only running certain linters

grunweg opened the pull request:
This adds a script `run_linter`, which is a fork of batteries' `runLinter` with extra features:
- rewrite as a CLI app, so --help messages etc. come for free
- add a `--list` option, which prints the list of linters it was about to run, and exists
- add `--select`, `--ignore` and `--only` options which ensure certain linters are always run, never resp. that only a certain list of linters are run. (`add` takes priority over `exclude` or `only`.)
- add an `update_only_remove` flag which only runs linters mentioned in the `nolints` file: this only removes entries (by design), but is often much faster.
- renamed to `run_linter`: actually, per the [zulip discussion](https://leanprover.zulipchat.com/#narrow/stream/287929-mathlib4/topic/Naming.20convention/near/450896930), this should be `run-linter` --- and one could as well use the opportunity to rename this to e.g. `env-linter`. Asked [here](https://leanprover.zulipchat.com/#narrow/stream/270676-lean4/topic/Configuring.2Fmodernizing.20runLinter).

Intended use cases
1. Run `update_only_remove` when updating the nolints file. In fact, it is so fast that it could even run during regular CI, after the normal `lint` step...
2. Run the mathlib linters locally. The simpNF linters are slow enough (for a reason, but still) that doing so is currently annoying. On my notebook, they can easily take half an hour, during which my notebook is otherwise hardly usable.
3. You're developing a new linter and only want to run *that* one (as there was no change about the other ones).

Probably, the nicest result would be this PR landing in batteries (perhaps in pieces) instead of here.

---

Two CLI warts I'm unhappy about.
- how can I have flag names with a hyphen? Currently, I have to write `--print_linters`, which is less pretty.
- many options take an `Array String`: how can I pass this to the executable? This is not obvious to me...

<!-- The text above the `---` will become the commit message when your
PR is merged. Please leave a blank newline before the `---`, otherwise
GitHub will format the text above it as a title.

For details on the "pull request lifecycle" in mathlib, please see:
https://leanprover-community.github.io/contribute/index.html

In particular, note that most reviewers will only notice your PR
if it passes the continuous integration checks.
Please ask for help on https://leanprover.zulipchat.com if needed.

To indicate co-authors, include lines at the bottom of the commit message
(that is, before the `---`) using the following format:

Co-authored-by: Author Name <author@email.com>

If you are moving or deleting declarations, please include these lines at the bottom of the commit message
(that is, before the `---`) using the following format:

Moves:
- Vector.* -> Mathlib.Vector.*
- ...

Deletions:
- Nat.bit1_add_bit1
- ...

Any other comments you want to keep out of the PR commit should go
below the `---`, and placed outside this HTML comment, or else they
will be invisible to reviewers.

If this PR depends on other PRs, please list them below this comment,
using the following format:
- [ ] depends on: #abc [optional extra text]
- [ ] depends on: #xyz [optional extra text]

-->

[![Open in Gitpod](https://gitpod.io/button/open-in-gitpod.svg)](https://gitpod.io/from-referrer/)

grunweg committed:
Boilerplate: copy-paste runLinter script from batteries
so I can edit this locally
Only minor changes: tweaked code style; changed default module to Mathlib.

grunweg committed:
refactor: convert myRunLinter to a CLI application

This will make adding further options easier, and gives e.g. --help output
for free. (It does add a dependency on the cli project, though.)

grunweg committed:
feat: implement the --only flag, only running certain linters

grunweg committed:
feat: also add --{add,exclude,print}_linters flags
Rename --only to --only_linters.

grunweg committed:
Copy-paste getCheck, so I can modify it.

grunweg committed:
Change up getChecks: remove the previous useOnly check,
which can be implemented easily as post-processing (and was broken anyway).
Replace by useAlso, which adds linter to always enable.
And use this to implement --add_linters

grunweg committed:
Be more permissive about my linter passing options:
--add is always allowed, and takes priority over --only or --exclude

grunweg committed:
feat: add --update-only-remove to only run the linters in the nolints file
And refactor linter selection once more: is complex, highly branched,
but at least very clear logic.

grunweg committed:
Rename to run_linter.

grunweg committed:
Tweak treatment of conflicting flags: allow both add and only
or add and exclude, but error if there are *actual* linters specified
which conflicts.

grunweg committed:
Rename CLI flags to {always,never,only}_run.

grunweg committed:
Avoid explicit path separator; use mkFilePath instead.

grunweg committed:
Rename useAlso -> runAlways.

grunweg committed:
refactor: move runOnly check back into getChecks

rosborn commented on scripts/run_linter.lean:
```diff
@@ -0,0 +1,167 @@
+  "Runs the linters on all declarations in the given module (or `Mathlib` by default)."
+
+  FLAGS:
+    only_linters : Array String;  "Only run these named linters"
+    exclude_linters : Array String; "Do not run these named linters"
+    add_linters : Array String; "Run these linters *in addition* to the default set\n\
+      Takes priority over the --only_linters or --exclude_linters flags."
+    print_linters; "Print the list of all discovered/configured linters and exit"
```
Remove `_linters` from these options. I would recommend using `--only`, `--ignore`, `--select`, and `--list` for these options (`include` and `exclude` are usually for files to lint or avoid linting). The priority should be on the last option provided. So `--select foo --ignore foo` would ignore, but `--ignore foo --select foo` would run `foo`.

grunweg replied:
Thanks for the feedback! I like your renaming suggestions, I'll do that in a moment.

I'm not sure regarding the order of options matter (my current instinct is to complain about contradictory information). Do you have any examples in mind which are order-sensitive --- or cases where specifying contradictory information is actually useful?

rosborn replied:
It is useful for when you create a shell alias `alias mylint="lake exe run_linter --only foo"` and then using the alias `mylint --select bar` will work as expected.

grunweg committed:
The PR to batteries has been merged.

grunweg committed:
Rename CLI flags as proposed.

grunweg committed:
Merge branch 'master' into MR-myRunLinter

grunweg committed:
chore(Topology/UniformSpace): make explicit variables on both sides of an iff implicit

A few errors remain, because of a bug in mk_iff --- this generates explicit
variable binders on both sides.

grunweg committed:
Fix Ascoli.lean

grunweg committed:
More fixes.

grunweg committed:
Fix and further.

grunweg committed:
Fix

grunweg committed:
Fix

grunweg committed:
Fix and further.

grunweg committed:
wip: Further and fixes until Topology/Perfect,
except for Germ and PartitionOfUnity whose deps don't build yet

grunweg committed:
Fix.

grunweg committed:
Changes to Separation

grunweg committed:
Fixes; revert one overzealous change.

grunweg committed:
temp

grunweg committed:
Small progress. TODO: need to revert or re-inspect the PartialHomeomorph changes

grunweg committed:
Merge branch 'master' into MR-myRunLinter
